### PR TITLE
Adapt product registration to the new architecture

### PR DESCRIPTION
**Pull request by jreidinger**
Another change in set that switch web UI from using dbus to HTTP API. As part of this change the rust code is added to provide that HTTP API and also adapted web UI. It also fixes all relevant js tests.

Things discussed but postponed (due to size of change and time constraints ) in this change was:

1. move registration requirement from registration to product as it is highly coupled with it
2. provide all changed data in websocket signal, so client does not need to do another http call to get correct data

Matching trello card: https://trello.com/c/2bTZOnpB

### CODE REVIEW

File: rust/agama-server/src/software/web.rs
Comment (imobachgs) [inline rust/agama-server/src/software/web.rs:66]: If we move the registration requirement to the product itself (after all, that's part of the product, right?), we do not need this signal.

File: rust/agama-server/src/software/web.rs
Comment (imobachgs) [inline rust/agama-server/src/software/web.rs:147]: So we need to ask for the registration information again, right? Perhaps it would be better to send the registration information along with the event.

File: rust/agama-lib/src/product/client.rs
Comment (jreidinger) [reply rust/agama-lib/src/product/client.rs:121]: problem with `?` is  that it is not ServiceError, so I will need to convert that error somehow ( doable as I own error ), but I think it is not worth effort.

File: rust/agama-server/src/software/web.rs
Comment (jreidinger) [reply rust/agama-server/src/software/web.rs:147]: well, it is possible, I will just need to recreate info again from proxy as I cannot get it from signal as it contain just one param and not all of them.

File: rust/agama-server/src/software/web.rs
Comment (jreidinger) [reply rust/agama-server/src/software/web.rs:66]: that is right, but value is computed in registration interface and not product one (on dbus side).

Comment (imobachgs) [reply rust/agama-server/src/software/web.rs:66]: JFTR, we agreed on postponing this change.

File: rust/agama-server/src/software/web.rs
Comment (imobachgs) [reply rust/agama-server/src/software/web.rs:148]: We already solve a problem like this with the [ProgressStream](https://github.com/openSUSE/agama/blob/architecture_2024/rust/agama-server/src/web/common.rs#L173-L220) API, that includes the full struct in the events. We can folllow a similar approach here.

However, we decided to keep it simple by now, but we have plans to improve it later.

**Comment by jreidinger** [comment]
I get this warning in FF using arch branch. @lslezak @imobachgs does it ring any bells to you?
```
Cookie “agamaToken” does not have a proper “SameSite” attribute value. Soon, cookies without the “SameSite” attribute or with an invalid value will be treated as “Lax”. This means that the cookie will no longer be sent in third-party contexts. If your application depends on this cookie being available in such contexts, please add the “SameSite=None“ attribute to it. To know more about the “SameSite“ attribute, read https://developer.mozilla.org/docs/Web/HTTP/Headers/Set-Cookie/SameSite
```

**Comment by imobachgs** [comment]
> I get this warning in FF using arch branch. @lslezak @imobachgs does it ring any bells to you?
> 
> ```
> Cookie “agamaToken” does not have a proper “SameSite” attribute value. Soon, cookies without the “SameSite” attribute or with an invalid value will be treated as “Lax”. This means that the cookie will no longer be sent in third-party contexts. If your application depends on this cookie being available in such contexts, please add the “SameSite=None“ attribute to it. To know more about the “SameSite“ attribute, read https://developer.mozilla.org/docs/Web/HTTP/Headers/Set-Cookie/SameSite
> ```

I would say we do not need to send the cookies in third-party contexts. We could easily set the `SameSite` value in the [atuh_cookie_from_token function](https://github.com/openSUSE/agama/blob/architecture_2024/rust/agama-server/src/web/http.rs#L132).

File: rust/agama-lib/src/product/client.rs
Comment (imobachgs) [inline rust/agama-lib/src/product/client.rs:27]:
```suggestion
    /// It is mandatory to register the product
```

File: rust/agama-server/src/software/web.rs
Comment (imobachgs) [inline rust/agama-server/src/software/web.rs:122]: As we agreed, I will create a card for this.

File: rust/agama-server/src/software/web.rs
Comment (imobachgs) [inline rust/agama-server/src/software/web.rs:277]: I would most probably use PUT, as you are replacing (or creating) the resource.

File: rust/agama-server/src/software/web.rs
Comment (imobachgs) [inline rust/agama-server/src/software/web.rs:238]: Why not use `Option<String>`?

File: rust/agama-server/src/software/web.rs
Comment (imobachgs) [inline rust/agama-server/src/software/web.rs:286]: Following your own suggestion, I would return `Ok(StatusCode::NO_CONTENT)` (you need to change the return type to `Result<impl IntoResponse, Error>`). We avoid then returning a useless `null`. The same for the `deregister` method below.

File: rust/agama-server/src/web/common.rs
Comment (imobachgs) [inline rust/agama-server/src/web/common.rs:20]: I like this but, what about using `EventsStream` as a more descriptive name?

File: web/src/Main.jsx
Comment (imobachgs) [inline web/src/Main.jsx:29]: :+1:

File: web/src/client/software.js
Comment (imobachgs) [inline web/src/client/software.js:94]: :+1:

File: web/src/client/software.js
Comment (imobachgs) [inline web/src/client/software.js:258]: No need to add a whitespace there, `console.log` takes care.

```suggestion
      console.log("Failed to get registration config:", response);
```

File: web/src/client/software.js
Comment (imobachgs) [inline web/src/client/software.js:263]: You can use destructuring to get the date like this:

```
    const { requirement, email, code } = config;
```

File: web/src/client/software.js
Comment (imobachgs) [inline web/src/client/software.js:286]: Why? Any chance we can fix this before merging?

File: web/src/client/software.js
Comment (imobachgs) [inline web/src/client/software.js:300]: What does it mean?

**Review by imobachgs — COMMENTED** [review]
In general, it looks good. Just a few suggestions.

File: rust/agama-server/src/software/web.rs
Comment (jreidinger) [reply rust/agama-server/src/software/web.rs:277]: well, if I use PUT, I will have to create it idempotent, which it is currently not.

File: rust/agama-server/src/software/web.rs
Comment (jreidinger) [reply rust/agama-server/src/software/web.rs:238]: I am considering it, but in the end I use same values as in dbus API. Reason is mainly to prevent accidental sending just email, that can mean that system is not registered.

File: rust/agama-server/src/software/web.rs
Comment (jreidinger) [reply rust/agama-server/src/software/web.rs:286]: well, for me reason is that POST has mandatory response body according to https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/POST

And that reminds me that what is still missing from old API is more detailed error codes. I am not sure if we can have something like `Ok(Json(scc_error))`?

File: rust/agama-server/src/web/common.rs
Comment (jreidinger) [reply rust/agama-server/src/web/common.rs:20]: in that case I would use `EventStreams` as it is not just single stream, but array of streams.

File: web/src/client/software.js
Comment (jreidinger) [reply web/src/client/software.js:263]: you are right, as conversion for requirement is no longer there, it is not needed

File: web/src/client/software.js
Comment (jreidinger) [reply web/src/client/software.js:300]: well, it calls delete. And if it failed we need to somehow get error code. I am not sure if way would be using body in Delete and 200 code?

File: rust/agama-server/src/software/web.rs
Comment (imobachgs) [reply rust/agama-server/src/software/web.rs:277]: Well, I am not sure. If you perform the same action with the same data several times, what changes? I would expect the result to be the same. But I am fine with using POST.

File: rust/agama-server/src/software/web.rs
Comment (imobachgs) [reply rust/agama-server/src/software/web.rs:238]: I still prefer using `Option<T>` that the typical "empty means null" handling. In D-Bus we are far more limited. Of course, I am not blocking the merge because of this, just mentioning.

File: rust/agama-server/src/web/common.rs
Comment (imobachgs) [reply rust/agama-server/src/web/common.rs:20]: Yes, you are right, that's even better.

File: rust/agama-server/src/software/web.rs
Comment (jreidinger) [reply rust/agama-server/src/software/web.rs:277]: well, the change will be that after first successful registration the next one fails. Also we do not handle if you use different data. You have to first deregister and then can register with different data. We handle that automatic only when you change product.

File: web/src/client/software.js
Comment (imobachgs) [reply web/src/client/software.js:300]: Sure, we need the error code. My question was more about why is this a TODO? If we cannot do it right now, then it is fine.

About the answer, I would use "422 Unprocessable Content" (perhaps there is a better code) and put the error code in the body.

**Comment by jreidinger** [comment]
JFYI testing of error response succeed:

registration_failure

**Review by imobachgs — APPROVED** [review]

**Pull request merged**
